### PR TITLE
Feature/bca/perf better send time

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Improve send time performances why removing usage of workers to send events.

On some devices heavy delay were experienced when sending events ( 5s, 6s, 20s...). It's touching some devices and not others and is still a bit random.

Workers by them selves are adding a bit of delay before launching, and in addition we have to rely on database to get up to date localEchos, and such transactions are sometime very slow

Example of improvement of send time (top is with worker, bottom is with new implementation)
image

The current implem just uses a sender thread, that consumes send tasks in order.
Media messages are still using workers

In addition some improvement have been made on timeline with broader support of UI local echos

** Known Limitation
 - The first message sent in an encrypted room will still take long to be sent, improvment for that is coming in another PR

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/DefaultSendService.kt [hunk 1]
```diff
@@ -93,17 +91,6 @@ internal class DefaultSendService @AssistedInject constructor(
     }
 
     // For test only
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/DefaultSendService.kt:93]: also remove this comment :)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/EventSenderProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+ * If the app is killed before all event were sent, on next wakeup the sheduled events will be re posted
+ */
+@SessionScope
+internal class EventSenderProcessor @Inject constructor(
+        private val cryptoService: CryptoService,
+        private val sessionParams: SessionParams,
+        private val queuedTaskFactory: QueuedTaskFactory,
+        private val taskExecutor: TaskExecutor,
+        private val memento: QueueMemento
+) : Thread("SENDER_THREAD_SID_${sessionParams.credentials.sessionId()}") {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/EventSenderProcessor.kt:58]: Have you tested the signout, if there are unsent Events?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/QueueMemento.kt [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * Simple lightweight persistence
+ * Don't want to go in DB due to current issues
+ * Will never manage lots of events, it simply uses sharedPreferences.
+ * It is just used to remember what events/localEchos was managed by the event sender in order to
+ * reschedule them (and only them) on next restart
+ */
+internal class QueueMemento @Inject constructor(context: Context,
+                                                sessionParams: SessionParams,
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/QueueMemento.kt:37]: You can inject @SessionId instead

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/TaskInfo.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import com.squareup.moshi.Moshi
+import org.matrix.android.sdk.api.extensions.tryOrNull
+import org.matrix.android.sdk.internal.di.SerializeNulls
+import org.matrix.android.sdk.internal.network.parsing.RuntimeJsonAdapterFactory
+
+/**
+ * Info that need to be persisted by the sender thread
+ * With polymorphic moshi parsing
+ */
+interface TaskInfo {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/TaskInfo.kt:30]: `internal` ?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/TaskInfo.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+@JsonClass(generateAdapter = true)
+data class RedactEventTaskInfo(
+        @Json(name = "type") override val type: String = TaskInfo.TYPE_REDACT,
+        @Json(name = "redactionLocalEcho") val redactionLocalEcho: String?,
+        @Json(name = "order") override val order: Int
+) : TaskInfo
+
+@JsonClass(generateAdapter = true)
+data class FallbackTaskInfo(
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/TaskInfo.kt:73]: all those 3 data classes should have `internal` modifier I think

**Review by bmarty — COMMENTED** [review]
I made a first static review. That's amazing work!
Will try to use the code during the week-end

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/QueueMemento.kt (hunk 3 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/QueueMemento.kt:37]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/TaskInfo.kt (hunk 4 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/TaskInfo.kt:30]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/TaskInfo.kt (hunk 5 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/TaskInfo.kt:73]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/EventSenderProcessor.kt (hunk 2 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/queue/EventSenderProcessor.kt:58]: When the session is closed, the send processor will be interrupted thus pending messages won't be sent.

**Comment by BillCarsonFr** [comment]
Fixes #516

**Pull request merged**